create a page for the workspace example contract (#706)
stellar/stellar-docs: docs/smart-contracts/example-contracts/workspace.mdx

Let's take a look at each crate, and see how they all work together.

### Contract A Interface: The Trait

The `contract_a_interface` crate defines a trait containing function footprint(s), which will need to be further fleshed out in the `contract_a` crate. This interface is defined here, separate from any business logic and only defines what functions should exist in a contract, as well as that function's parameter(s) and return type(s).

You can see our interface defines `add` as the only function our contract will contain. This `add` function, then requires two inputs, `x` and `y`, both of which will be `u32` integers. Finally, the function should return a `u32` integer as well.

The use of `contractclient` as an attribute macro on `ContractAInterface` means a client will be created, conforming to this interface, that can be used by contracts existing outside of the `contract_a_interface` crate. As you'll see later, we use this client, `ContractAClient`, in the `contract_b` crate to call the `add` function.

### Contract A: The Logic

> If the `contract_a_interface` crate is only present to define which functions, inputs, and outputs should be part of our contract, the `contract_a` crate is only present to actually code what the interface function(s) should actually _do_ to return the correct type of response.

Reviewer: ```suggestion
The `contract_a` crate contains the implementation for Contract A, defining for each function what it should actually _do_.
```

> You can see our contract implementation of the `ContractAInterface` takes the value of `x`, along with the value of `y`, and performs a `checked_add`, returning the sum of both numbers (while avoiding an overflow error).

Reviewer: ```suggestion
The implementation takes the value of `x`, along with the value of `y`, and performs a `checked_add`, returning the sum of both numbers (while avoiding an overflow error).
```

:::info

> You may have noticed this contract doesn't require an `Env` argument. That's totally fine! `Env` has a ton of useful features that are available to your smart contracts, if you need them. But, you're not at all required to use it, if you don't.

Reviewer: ```suggestion
The `add` function doesn't require an `Env` argument. That's totally fine! `Env` has many useful features that are available to your smart contracts, if you need them. But, you're not at all required to use it, if you don't.
```

:::

> All that is required to make use of the previously defined `ContractAInterface` is to `use` the interface, define a `ContractA` struct, and then `impl` the interface on top of that struct.

Reviewer: ```suggestion
All that is required to make use of the previously defined `ContractAInterface` is to `use` the interface, define a `ContractA` struct, and `impl` the interface for that struct.
```

This crate uses the `contractimpl` attribute macro on the `ContractA` implementation, making the `add` function public and invocable by others on the Stellar network.

### Contract B: The Invocation

> Now that we've created a trait in `contract_a_interface`, and implemented it in `contract_a`, we can use the `contract_b` crate to actually invoke the `add` function and get the sum of our integers.

Reviewer: ```suggestion
Now that we've created a trait in `contract_a_interface`, and implemented it in `contract_a`, we can use the `contract_b` crate to invoke the `add` function and get the sum of our integers.
```

> We're creating and implementing an entirely new contract, `ContractB`. We're skipping the trait, and getting right to the contract itself. We're making a function, `add_with`, which will run inside the Soroban environment. When it is invoked, it requires three arguments:

Reviewer: ```suggestion
We're creating and implementing an entirely new contract, `ContractB`. We're skipping the trait, and getting right to the contract itself.

::: info

Defining the contract interface with a `trait` separately is optional. It's a great way to share the interface of a contract, and a contract client, with multiple contracts in a multi-contract workspace.

:::

`ContractB` has one function, `add_with`, that requires three arguments:
```

> - `contract: Address` - The address of a contract which implements that required client interface, `ContractAClient`.

Reviewer: ```suggestion
- `contract: Address` - The address of a contract which implements that required client interface, `ContractAInterface`.
```

- `x: u32` and `y: u32` - The two numbers we want to (safely) compute the sum of.

> `ContractB` then invokes the `add` function from the deployed `ContractA`, returning its value back to the original invoker. It's a bit of a long round-trip for this simple example, but it illustrates a really powerful way you can separate out interface/trait definitions from contract logic in a way that allows crates to be reusable in a very powerful way.

Reviewer: ```suggestion
`ContractB` invokes `ContractA`'s `add` function, returning its value back to the original invoker. It's a bit of a long round-trip for this simple example, but it illustrates a really powerful way you can separate out interface/trait definitions from contract logic and share it and its client in a multi-contract workspace.
```

## Practical Use-Case Examples

> Outside the world of adding two integers together, this technique is far more versatile and useful in the real world. For example, this strategy could be used to:

Reviewer: ```suggestion
Beyond this simple example, this technique is versatile and useful. For example, this strategy could be used to:
```

> - Create and reference a standardized, consistent token interface?
> - Reuse a single interface that you want to incorporate across many different contracts?
> - Other things, too?

Reviewer: We should say what the good uses are, and avoid the etc.
```suggestion
- Create and reference a standardized, consistent token interface.
- Reuse a single interface that you want to incorporate across many different contracts.
```

## Build the Contracts

> To build the contract into a set of `.wasm` files, use the `stellar contract build` command. Both `workspace/contract_a` and `workspace/contract_b` must be built, and you can use a single command, since our workspace defines its `members` in the `Cargo.toml` file:

Reviewer: ```suggestion
To build the contracts into a set of `.wasm` files, use the `stellar contract build` command. Both `workspace/contract_a` and `workspace/contract_b` will be built, and you can use a single command, since our workspace defines its `members` in the `Cargo.toml` file:
```

```shell
stellar contract build
```

Two `.wasm` files should be found in the `workspace/target` directory:

```text
target/wasm32-unknown-unknown/release/soroban_workspace_contract_a.wasm
target/wasm32-unknown-unknown/release/soroban_workspace_contract_b.wasm
```

> Since the `contract_a_interface` doesn't make use of the `contract` attribute macro, the Stellar CLI knows there's nothing to build, so it doesn't even bother. Nice!

Reviewer: ```suggestion
The [`stellar-cli`] knows the `contract_a_interface` is not intended to be compiled into a .wasm, because the `contract_a_interface`'s `Cargo.toml` has its `crate-type` configured as `rlib` (rust library). Nice!
```
